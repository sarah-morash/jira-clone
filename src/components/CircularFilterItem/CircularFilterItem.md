###

Click Me!

```
initialState={active: false};
<CircularFilterItem selected={state.active} filterText="Filter" onClickFunction={() => setState({active: !state.active})} />
```
