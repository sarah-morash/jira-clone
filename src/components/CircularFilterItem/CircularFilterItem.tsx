import { Button } from "@material-ui/core";
import * as React from "react";

export interface ICircularFilterItem {
  /**
   * The id of this filter item
   */
  id: string;

  /**
   * Determining if this item
   * is currently selected
   */
  selected?: boolean;

  /**
   * The test ot appear within
   * this filter item
   */
  filterText: string;

  /**
   * Function to be called
   * when the item is clicked
   * upon
   */
  onClickFunction: (filterId: string) => void;
}

export default (props: ICircularFilterItem) => (
  <Button
    variant={props.selected ? "contained" : "text"}
    color="primary"
    onClick={() => props.onClickFunction(props.id)}>
    {props.filterText}
  </Button>
);
