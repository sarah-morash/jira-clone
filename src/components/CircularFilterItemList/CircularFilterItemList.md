```
initialState = {
  filters: [
    {
      selected: false,
      id: "abc",
      filterText: "Filter 1"
    },
    {
      selected: false,
      id: "def",
      filterText: "Filter 2"
    }
  ]
};

<CircularFilterItemList filters={state.filters}
  onChangeFilterFunction={(id) =>
  setState( {filters: state.filters.map( filter => {
    if (filter.id === id)
    {
      return Object.assign({}, filter, {selected: !filter.selected})
    }
    return filter;
  })})}
/>
```
