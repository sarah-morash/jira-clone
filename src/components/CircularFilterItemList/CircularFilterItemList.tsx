import * as React from "react";
import styled from "react-emotion";
import CircularFilterItem from "../CircularFilterItem/CircularFilterItem";

export interface IFilter {
  id: string;
  selected: boolean;
  filterText: string;
}

interface ICircularFilterItemList {
  /**
   * List of filters
   */
  filters: IFilter[];

  /**
   * Function to be invoked
   * when the user clickes a
   * filter
   */
  onChangeFilterFunction: (clickedFilterId: string) => void;
}

const CircularFilterItemWrapper = styled.div`
  display: inline-block;
  &:nth-child(n + 2) {
    padding-left: 5px;
  }
`;
export default (props: ICircularFilterItemList) => (
  <div>
    {props.filters.map((filter: IFilter) => (
      <CircularFilterItemWrapper key={filter.id}>
        <CircularFilterItem
          {...filter}
          onClickFunction={(filterId: string) =>
            props.onChangeFilterFunction(filterId)
          }
        />
      </CircularFilterItemWrapper>
    ))}
  </div>
);
