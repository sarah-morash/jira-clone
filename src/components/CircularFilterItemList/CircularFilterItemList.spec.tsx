import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import * as sinon from "sinon";
import CircularFilterItem from "../CircularFilterItem/CircularFilterItem";
import CircularFilterItemList, { IFilter } from "./CircularFilterItemList";

describe("Circular Filter List", () => {
  let wrapper: ShallowWrapper;
  const sampleFilters: IFilter[] = [
    {
      selected: false,
      id: "abc",
      filterText: "Filter 1"
    },
    {
      selected: false,
      id: "def",
      filterText: "Filter 2"
    }
  ];
  beforeEach(() => {
    wrapper = shallow(
      <CircularFilterItemList
        filters={sampleFilters}
        onChangeFilterFunction={() => {
          return;
        }}
      />
    );
  });

  /**
   * Snapshot
   */
  it("Should match its snapshot", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  /**
   * Rendering List of Items
   */
  it("Should render a <CircularFilterItem /> for each item within its filters prop", () => {
    expect(wrapper.find(CircularFilterItem)).toHaveLength(2);
  });

  /**
   * Passing its callback to each item
   */
  it("Should call its onFilterChangeFunction function when the user clicks upon a filter", () => {
    const sampleFunction = sinon.spy();
    const mounted = mount(
      <CircularFilterItemList
        filters={sampleFilters}
        onChangeFilterFunction={sampleFunction}
      />
    );
    mounted
      .find(CircularFilterItem)
      .first()
      .simulate("click");
    expect(sampleFunction).toHaveProperty("callCount", 1);
  });

  it("Should call its onFilterChangeFunction function with the id of the filter clicked", () => {
    const sampleFunction = sinon.spy();
    const mounted = mount(
      <CircularFilterItemList
        filters={sampleFilters}
        onChangeFilterFunction={sampleFunction}
      />
    );
    mounted
      .find(CircularFilterItem)
      .first()
      .simulate("click");
    expect(sampleFunction.getCall(0).args[0]).toMatch(sampleFilters[0].id);
  });
});
