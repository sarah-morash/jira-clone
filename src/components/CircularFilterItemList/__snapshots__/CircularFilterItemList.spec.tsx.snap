// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circular Filter List Should match its snapshot 1`] = `
<div>
  <CircularFilterItemWrapper
    key="abc"
  >
    <Component
      filterText="Filter 1"
      id="abc"
      onClickFunction={[Function]}
      selected={false}
    />
  </CircularFilterItemWrapper>
  <CircularFilterItemWrapper
    key="def"
  >
    <Component
      filterText="Filter 2"
      id="def"
      onClickFunction={[Function]}
      selected={false}
    />
  </CircularFilterItemWrapper>
</div>
`;
